src/saga_exec.rs

```diff
@@ -257,6 +281,28 @@ impl<UserType: SagaType> SagaNodeRest<UserType> for SagaNode<SgnsUndone> {
     }
 }
 
+impl<UserType: SagaType> SagaNodeRest<UserType> for SagaNode<SgnsUndoFailed> {
+    fn log_event(&self) -> SagaNodeEventType {
+        SagaNodeEventType::UndoFailed(self.state.0.clone())
+    }
+
+    fn propagate(
+        &self,
+        _exec: &SagaExecutor<UserType>,
+        live_state: &mut SagaExecLiveState,
+    ) {
+        assert!(live_state.exec_state == SagaCachedState::Unwinding);
+        assert!(!live_state.undo_errors.contains_key(&self.node_id));
+        live_state
+            .undo_errors
+            .insert(self.node_id, self.state.0.clone())
+            .expect_none("undo node failed twice (storing error)");
```

Comment:
NIt: The `assert(!live_state.undo_errors.contains_key(&self.node_id))` above is an equivalent assertion. I'd probably remove the former.